Extend debug utilities for stuck/errored tracks

sethrj opened the pull request:
While debugging a hang in ATLAS, we tried the `debug_print` method [documented in our manual](https://celeritas-project.github.io/celeritas/user/development/testing.html) to see track information, but the symbol wasn't available due to use of LTO and/or `-Wl,--exclude-libs,ALL` and/or static libs. By adding a secret environment-based call, which we've placed in the most sensible location of `KernelContextException` to ensure that symbol stays in the library.

This also adds a `kill_active` method to the `Stepper` that can be called in a debugger or in between steps. It marks all tracks as "errored" and will apply a tracking cut at the next step, which (on CPU) also prints diagnostic information about the track.

A new `SetupOptions::geometry_output_file` can be used to export the in-memory geometry alongside the physics and primaries to aid in reproducing errors.

Finally, this greatly enhances the diagnostic output from errored/cut tracks, now printing out a JSON dump of the track *with* names substituted via core params metadata. This is accomplished by hiding a host-only "observer pointer" to CoreParams inside the core params data. Because that pointer is only usable on the host and only needed inside kernels, it shouldn't result in people abusing it.
```
Killing track {"geo":{"dir":[1.0,0.0,0.0],"is_on_boundary":true,"is_outside":false,"pos":[[-5.0,0.0,0.0],"cm"],"volume_id":"inner@0x0"},"particle":{"energy":[[0.0,"MeV"],"MeV"],"particle_id":"gamma"},"sim":{"event_id":0,"num_steps":1,"parent_id":-1,"post_step_action":"tracking-cut","status":"errored","step_length":[17.0,"cm"],"time":[0.25,"s"],"track_id":1},"thread_id":7,"track_slot_id":7}: lost 100 MeV
```

#1144

sethrj committed:
Add a reference to the debug_print symbol

pcanal commented on src/celeritas/global/KernelContextException.cc:
```diff
@@ -133,6 +135,16 @@ void KernelContextException::initialize(CoreTrackView const& core)
             surface_ = geo.surface_id();
         }
+
+        if (!celeritas::getenv("CELERITAS_EXCEPTION_DEBUG").empty())
```
When is `void KernelContextException::initialize(CoreTrackView const& core)` called?

sethrj replied:
During the constructor of the KernelContextException... which happens if there's an assertion/runtime failure inside one of the `launch_action` loops.

sethrj committed:
Change variable name for consistency

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into debug-print-export

sethrj committed:
Add helper function to kill active tracks

sethrj committed:
Add test

sethrj committed:
Add option to export geant geometry at runtime

sethrj committed:
Write geometry file in example

sethrj committed:
Fix cuda build

sethrj committed:
Fix stepper test

sethrj committed:
Improve accel SetupOptions docs

sethrj committed:
Add geant UI for geometry export option

drbenmorgan commented:
This looks o.k. to me modulo the tests passing!

pcanal commented on test/celeritas/global/Stepper.test.cc:
```diff
@@ -229,6 +226,75 @@ TEST_F(SimpleComptonTest, TEST_IF_CELER_DEVICE(device))
+    {
+        static char const* const expected_log_messages[] = {
+            "Killing 2 active tracks",
+            R"(Killing track 0 of event 0 (in track slot 6) at {-5, 0, 0} cm along {1, 0, 0}: depositing 100 MeV in volume 1)",
+            R"(Killing track 1 of event 0 (in track slot 7) at {-5, 0, 0} cm along {1, 0, 0}: depositing 100 MeV in volume 1)"};
```
```suggestion
            R"(Killing track 0 of event 0 (in track slot 6) at {-5, 0, 0} cm along {1, 0, 0}: depositing 100 MeV in volume 0)",
            R"(Killing track 1 of event 0 (in track slot 7) at {-5, 0, 0} cm along {1, 0, 0}: depositing 100 MeV in volume 0)"};
```

sethrj replied:
Unfortunately the reverse happens on my machine... hence the update to actually print the volume *names* since the IDs might be nondeterminstic...

sethrj committed:
Improve Observerptr semantics

sethrj committed:
Add a global params pointer for detailed debug information

This piggybacks a pointer to the CoreParams object for use in the debug
output, which can now be constructed as a "streamable" object.

sethrj committed:
Fix unit test failure from using '@global' in JSON ORANGE

sethrj committed:
Be even pickier

sethrj committed:
Add documentation
